services/hack/submitHack.js

```diff
@@ -2,24 +2,21 @@ import AWS from "aws-sdk";
 import { pickIfTruthy } from "../../utils";
 import { connectToDatabase } from "../database/db";
 import Submission from "../database/models/SubmissionModel";
+import mongoose from "mongoose";
 
 AWS.config.update({ region: "us-east-1" });
-const s3 = new AWS.S3({ apiVersion: "2006-03-01" });
-
-// Call S3 to list the buckets
-s3.listBuckets(function(err, data) {
-  if (err) {
-    console.log("Error", err);
-  } else {
-    console.log("Success", data.Buckets);
-  }
-});
 export const submit = async (event, context) => {
   context.callbackWaitsForEmptyEventLoop = false;
   const data = JSON.parse(event.body);
-  let { submissionId } = data;
-
+  let { submissionId, files } = data;
+  if (submissionId === null){
+    submissionId = mongoose.Types.ObjectId();
+  }
   const submit = pickIfTruthy(data, "hackId", "message");
+  submit.files = [];
+  for (let key in files){
+    submit.files.push({name: files[key].name, size: files[key].size, type: files[key].type});
+  }
```

Comment:
I think it would be possible to have orphaned files in s3. The issue I have come across with uploading files to lambda than to s3 is the large increase in cpu time that would come about with uploading files directly to the lambda then to the s3. The advice I can find is to upload to s3 directly.

Would a possible solution and one that would fix you comment below would be to attach the hackID to the filename? which would allow us to scan through s3 (not sure if this could be a scaling issue)
so the process could be 
1. click submit on hack details
2. check files in s3 to see if there are any orphaned files and add them to the submission
click on submit button on submit page
3. Upload files to lambda
4. add submission data to database

Reply:
as part of getting the upload url, can we update the hack to know about the file? that way the file wont get orphaned in s3, you'd just have to handle the hack having files that may never have made it into s3 (which you'll always have to do).

crazy idea, can we create a bucket per hack, named after the hack id, so all files uploaded to it are for the one hack (avoiding naming conflicts) and no need to store the files against the hack at all.

Reply:
just for the record you [can create buckets](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/s3-example-creating-buckets.html#s3-example-creating-buckets-new-bucket) and [list the files from it](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/s3-example-creating-buckets.html#s3-example-listing-objects).

Not sure its a good idea, but [pricing](https://aws.amazon.com/s3/pricing/) does appear to change based on the number of buckets you have...

Reply:
So we would initialise the submission document in the database with the data then upload the files then update the submission document again with the download link for each file?

I think from a cost perspective It would be best to stick with one bucket per hack.

Reply:
The new current working logic is:
1. Click Submit
2. files and details saved to database
3. file upload begins (uploading to a directory for that particular hack idea)
4. Once upload has completed double check the s3 and database have the same files and remove any files that are not in the database